claat/parser/md/parse.go

```diff
@@ -0,0 +1,507 @@
+// limitations under the License.
+
+//Package mg implements a parser for CLaaT. It expects, as input, the output of running a Markdown file through
+//the Devsite Markdown processor.
+package md
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/net/html"
+	"golang.org/x/net/html/atom"
+
+	"github.com/googlecodelabs/tools/claat/parser"
+	"github.com/googlecodelabs/tools/claat/types"
+	"github.com/russross/blackfriday"
+)
+
+const (
+	metaSummary          = "summary"
+	metaID               = "id"
+	metaCategories       = "categories"
+	metaEnvironments     = "environments"
+	metaStatus           = "status"
+	metaFeedbackLink     = "feedback link"
+	metaAnalyticsAccount = "analytics account"
+)
+
+var metadataRegexp = regexp.MustCompile(`(.+?):(.+)`)
+var languageRegexp = regexp.MustCompile(`language-(.+)`)
+var durationHintRegexp = regexp.MustCompile(`(?i)Duration:? (.+)`)
+var durationRegexp = regexp.MustCompile(`(\d+)[:.](\d{2})$`)
+var downloadButtonRegexp = regexp.MustCompile(`^(?i)Download(.+)$`)
+
+// init registers this parser so it is available to CLaaT.
+func init() {
+	parser.Register("md", Parse)
+}
+
+// parserState encapsulates the state of the parser at any given step.
+type parserState struct {
+	tzr *html.Tokenizer
+	c   *types.Codelab
+	t   html.Token
+
+	currentStep *types.Step
+}
+
+// emit accepts a node, and either writes the node directly to the current step, or writes the node to the node buffer.
+func (ps *parserState) emit(n types.Node) {
+	ps.currentStep.Content.Append(n)
+}
+
+// advance moves the tokenizer to the next token and updates the token convenience variable.
+func (ps *parserState) advance() {
+	ps.tzr.Next()
+	ps.t = ps.tzr.Token()
+}
+
+// multiAdvance is a convenience method for repeatedly advancing.
+func (ps *parserState) multiAdvance(n int) {
+	for i := 0; i < n; i++ {
+		ps.advance()
+	}
+}
+
+// claatMarkdown calls the Blackfriday Markdown parser with some special addons selected. It takes a byte slice as a parameter,
+// and returns its result as a byte slice.
+func claatMarkdown(b []byte) []byte {
+	extns := blackfriday.EXTENSION_FENCED_CODE |
+		blackfriday.EXTENSION_NO_EMPTY_LINE_BEFORE_BLOCK |
+		blackfriday.EXTENSION_NO_INTRA_EMPHASIS |
+		blackfriday.EXTENSION_DEFINITION_LISTS |
+		blackfriday.EXTENSION_TABLES
+	o := blackfriday.Options{
+		Extensions: extns,
+	}
+	r := blackfriday.HtmlRenderer(extns, "", "")
+	return blackfriday.MarkdownOptions(b, r, o)
+}
+
+// Parse is a parser.ParseFunc for Markdown.
+func Parse(r io.Reader) (*types.Codelab, error) {
+	// Convert Markdown to HTML for easy parsing.
+	mBuf := bytes.Buffer{}
+	_, err := mBuf.ReadFrom(r)
+	if err != nil {
+		return nil, err
+	}
+	hBytes := claatMarkdown(mBuf.Bytes())
+	h := bytes.NewBuffer(hBytes)
+	// Parse the markup.
+	c, err := parseMarkup(h)
+	if err != nil {
+		return nil, err
+	}
+	return c, nil
```

Comment:
`return parseMarkup(h)` - no need for redundant err check if it's the last thing you do here anyway.